src/MooLite/plugins/EnhancerHelper/EnhancerHelperPlugin.ts

```diff
@@ -0,0 +1,213 @@
+        );
+        if (index != -1) {
+            return allItems[index];
+        }
+        return null;
+    }
+
+    getEnhancingToolBonus(item: CharacterItem | null): number {
+        if (item == null) {
+            return 0;
+        } else {
+            const enhancementLevel = item.enhancementLevel;
+            const itemDetails = this._game.inventory.itemDetailMap[item.itemHrid];
+            return this._game.enhancing.getEnhancingToolBonus(enhancementLevel, itemDetails);
+        }
+    }
+
+    getProtectionAmount(
+        protectionAmountTable: number[],
+        targetItemLevel: number,
+        currentEnhancementLevel: number
+    ): number {
+        const useProtection = <HTMLInputElement>document.getElementById("useProtect");
+        if (useProtection.checked) {
+            return protectionAmountTable[targetItemLevel] - protectionAmountTable[currentEnhancementLevel];
+        }
+        return 0;
+    }
+
+    getActionAmount(
+        actionAmountTable: number[],
+        enhancementCostTable: number[],
+        blessedTeaTable: number[],
+        targetItemLevel: number,
+        currentEnhancementLevel: number,
+        totalMaterialCost: number
+    ): number {
+        const useProtection = <HTMLInputElement>document.getElementById("useProtect");
+        const useBlessedTea = <HTMLInputElement>document.getElementById("blessedTea");
+        let actionAmount = 0;
+        if (useProtection.checked) {
+            actionAmount = actionAmountTable[targetItemLevel] - actionAmountTable[currentEnhancementLevel];
+        } else {
+            actionAmount =
+                (enhancementCostTable[targetItemLevel] - enhancementCostTable[currentEnhancementLevel]) /
+                totalMaterialCost;
+        }
+        if (useBlessedTea.checked) {
+            return actionAmount / blessedTeaTable[targetItemLevel];
+        }
+        return actionAmount;
+    }
+
+    getExperience(
+        sTable: number[],
+        actionAmount: number,
+        itemLevel: number,
+        currentLevel: number,
+        targetLevel: number
+    ): number {
+        const wisdomTea = <HTMLInputElement>document.getElementById("wisdomTea");
+        let experience = 0;
+        const baseExpRate = 1.4 * itemLevel + 14;
+        let remainingActionAmount = actionAmount;
+        for (let i = currentLevel + 1; i < targetLevel; i++) {
+            experience +=
+                baseExpRate * i * actionAmount * sTable[i] +
+                baseExpRate * i * (remainingActionAmount - actionAmount * sTable[i]) * (1 - sTable[i]) * 0.1;
+            remainingActionAmount = actionAmount * sTable[i];
+        }
+        experience += baseExpRate * targetLevel;
+        if (remainingActionAmount - 1 > 0) {
+            experience += baseExpRate * targetLevel * (remainingActionAmount - 1) * 0.1;
+        }
+        if (wisdomTea.checked) {
+            experience *= 1.15;
```

Comment:
Hardcoded values

Reply:
Fixed the item buffs but the experience equations aren't found in the init client info so no real choice but to hardcode it.